app/src/ble.c

```diff
@@ -447,12 +447,6 @@ static void connected(struct bt_conn *conn, uint8_t err) {
 
     LOG_DBG("Connected %s", addr);
 
-#if !IS_ENABLED(CONFIG_BT_GATT_AUTO_SEC_REQ)
-    if (bt_conn_set_security(conn, BT_SECURITY_L2)) {
-        LOG_ERR("Failed to set security");
-    }
-#endif // !IS_ENABLED(CONFIG_BT_GATT_AUTO_SEC_REQ)
-
```

Comment:
> So, we want to not only keep that setting off, we want to remove our explicit `bt_conn_set_security` in favor of having the GATT/ATT _built in_ security enforcement preventing connected centrals from accessing things before our security is established.

1. I assume, Central initiating pairing is checked for in the application for being in bondable state, right? or we just accept new bonds all the time from new centrals?
2. When connected to a bonded peer, is it ensured that notifications and indications have the required security level? where in the implementation is the request to encrypt initiated from the peripheral?

Reply:
@cvinayak

1. So, whenever we advertise, and the selected profile is "open" we accept pairing requests. You can see that in our `auth_pairing_accept` implementation:

```
static enum bt_security_err auth_pairing_accept(struct bt_conn *conn,
                                                const struct bt_conn_pairing_feat *const feat) {
    struct bt_conn_info info;
    bt_conn_get_info(conn, &info);

    LOG_DBG("role %d, open? %s", info.role, zmk_ble_active_profile_is_open() ? "yes" : "no");
    if (info.role == BT_CONN_ROLE_PERIPHERAL && !pairing_allowed_for_current_profile(conn)) {
        LOG_WRN("Rejecting pairing request to taken profile %d", active_profile);
        return BT_SECURITY_ERR_PAIR_NOT_ALLOWED;
    }

    return BT_SECURITY_ERR_SUCCESS;
};
```

2. We don't explicitly request encryption when trying to notify, but the Zephyr GATT code in `gatt_notify` (in `subsystem/bluetooth/host/gatt.c`) has a permission check that causes it to return `-EPERM` if our target connection is not encrypted yet. That should keep us from accidentally notifying if we don't yet have our connection encrypted. That being said, I know that *most* hosts will automatically upgrade their connection when re-connecting to previously bonded peers, but it *might* be necessary to have our code request encryption on notification as a fallback. Thoughts?

Reply:
> whenever we advertise, and the selected profile is "open" we accept pairing requests. You can see that in our auth_pairing_accept implementation:

This is ok then.

> it might be necessary to have our code request encryption on notification as a fallback. Thoughts?

Yes. Maybe you could check macOS/iOS (i do not remember on top of my mind) when bond is removed, and you connect back to a keyboard that has the previous bond and press keys, then what is the behavior? Now sending a SMP Security Request PDU may be of help?